### PR TITLE
JS: support additional flow syntax

### PR DESCRIPTION
**Pull request by esben-semmle**
This change enables the extractor to parse additional Flow syntax.

The commits are grouped by topic:

-   imports: support for `import` in declared modules, and support for `import type ...;`
-   exports: support for `export type * ...;`
-   type arguments: support for `f<T>(v)` (ODASA-7586)
    -   note that this syntax is ambiguous: this would ordinarily be parsed as `f < T > v`: we now prefer to parse it as a call, and use backtracking if that fails

A full evaluation is ongoing.

### CODE REVIEW

**Comment by esben-semmle** [comment]
Change notes added.
I have also rebased on `next` and retargeted (as recommended on Slack).
The rebase introduced a minor conflict and an opportunity for a refactoring - please have a look at `JS: support additional import statements for Flow`.

**Comment by xiemaisi** [comment]
How is the evaluation coming along?

**Comment by esben-semmle** [comment]
I have improved on the parsing of `f<T>(v)`. 
It turns out that minified asm.js-like code can be of the above form, causing a misparse. Examples: `if(!(0<(jd|0)&0>(Wc|0))){}`, `g=(g|0)<(e|0)|(g|0)>(b|0)?e:g;`

I have adjusted the parser to take care of those situations by checking if the parsed callee syntactically is a primitive from a string or number literal, or a unary or binary expression, disregarding parentheses. It is a bit more general than what two examples above require, but I believe it is a safe generalization.

The original evaluation without the above fix shows practically unchanged performance: https://git.semmle.com/esben/dist-compare-reports/tree/js/more-flow-parsing_1546739626480.
The recent sanity check evaluation shows that the fix eliminates the bad parses: https://git.semmle.com/esben/dist-compare-reports/tree/js/more-flow-parsing_1547044911097.

Please let me know if you agree with the fix before I start another full evaluation.

File: javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java [hunk 1]
```diff
@@ -1203,4 +1207,49 @@ protected boolean atGetterSetterName(PropertyInfo pi) {
...
+
+	private void maybeFlowParseTypeParameterInstantiation(Expression left, boolean requireParenL) {
+		if (flow() && this.isRelational("<")) {
+			// Ambiguous case: `e1<e2>(e3)` is parsed differently as JS and Flow code:
+			// JS: two relational comparisons: `e1 < e2 > e3`
+			// Flow: a call `e1(e3)` with explicit type parameter `e2`
+
+			// Heuristic: if the left operand of the `<` token is a primitive, then it probably isn't a call, as that would always crash
+			left = left.stripParens();
+			if (left instanceof Literal) {
```
Comment (xiemaisi) [inline javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java]: Are there any literals for which we would _not_ want to choose the JS parse?

Comment (esben-semmle) [reply javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java]: function literals?

Comment (xiemaisi) [reply javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java]: Those would be `FunctionExpression`s, no?

Comment (esben-semmle) [reply javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java]: I see. See the most recent commit: we now use the JS parse for all literals.

Comment (xiemaisi) [reply javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java]: LGTM, good to proceed with the full evaluation as far as I'm concerned.

**Comment by esben-semmle** [comment]
[Full evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/more-flow-parsing_1547408773607): 

- practically unchanged performance
- lots of result changes in test code due to the elimination of syntax error alerts which previously blocked many semantic alerts

This is good to go.

File: javascript/extractor/src/com/semmle/js/extractor/Main.java [hunk 2]
```diff
@@ -41,7 +41,7 @@
 	 * such a way that it may produce different tuples for the same file under the same
 	 * {@link ExtractorConfig}.
 	 */
-	public static final String EXTRACTOR_VERSION = "2018-12-19";
+	public static final String EXTRACTOR_VERSION = "2019-09-01";
```
Comment (xiemaisi) [inline javascript/extractor/src/com/semmle/js/extractor/Main.java]: This should have been `2019-01-09` for consistency, but it's just a tag, so doesn't matter.

**Pull request merged**